File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0594_Longest_Harmonious_Subsequence.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.HashMap;
import java.util.Map;

public class Question_0594_Longest_Harmonious_Subsequence {
    public int findLHS(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        Map<Integer, Integer> map = new HashMap<>();
        int res = 0;
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        for (int key : map.keySet()) {
            if (map.containsKey(key + 1)) {
                res = Math.max(res, map.get(key) + map.get(key + 1));
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0594_Longest_Harmonious_Subsequence.cc
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  int findLHS(vector<int>& nums) {
    if (nums.empty()) {
      return 0;
    }
    unordered_map<int, int> m;
    for (int num : nums) {
      m[num] += 1;
    }
    int res = 0;
    for (auto it = m.begin(); it != m.end(); it++) {
      if (m.count(it->first + 1) != 0) {
        res = max(res, it->second + m[it->first + 1]);
      }
    }
    return res;
  }
};
